pkg/core/resource_graph.go

```diff
@@ -235,3 +259,214 @@ func (rg *ResourceGraph) getAllDownstreamResourcesSet(source Resource, upstreams
+	if currValue != nil {
+		return currValue.(T), nil
+	}
+	return castedRes.(T), nil
+}
+
+func (rg *ResourceGraph) actOnValue(targetValue reflect.Value, res Resource, metadata any, parent *reflect.Value, index reflect.Value) error {
+	switch value := targetValue.Interface().(type) {
+	case Resource:
+		if targetValue.IsNil() {
+			value = reflect.New(targetValue.Type().Elem()).Interface().(Resource)
+		}
+		err := rg.callCreate(reflect.ValueOf(value), metadata)
+		currValue := rg.GetResourceByVertexId(value.Id().String())
+		if currValue != nil {
+			value = currValue
+		}
+		if err == nil && value != nil {
+			if parent != nil {
+				parent.SetMapIndex(index, reflect.ValueOf(value))
+			} else {
+				targetValue.Set(reflect.ValueOf(value))
+			}
+		} else {
+			return err
+		}
+	case *IaCValue:
+		if value != nil && value.Resource != nil {
+			err := rg.callCreate(reflect.ValueOf(value.Resource), metadata)
+			currValue := rg.GetResourceByVertexId(value.Resource.Id().String())
+			if currValue != nil {
+				value.Resource = currValue
+			}
+			if err == nil && value.Resource != nil {
+				if parent != nil {
+					parent.SetMapIndex(index, reflect.ValueOf(value))
+				} else {
+					targetValue.Set(reflect.ValueOf(value))
+				}
+			} else {
+				return err
+			}
+		}
+	case IaCValue:
+		if value.Resource != nil {
+			err := rg.callCreate(reflect.ValueOf(value.Resource), metadata)
+			currValue := rg.GetResourceByVertexId(value.Resource.Id().String())
+			if currValue != nil {
+				value.Resource = currValue
+			}
+			if err == nil && value.Resource != nil {
+				if parent != nil {
+					parent.SetMapIndex(index, reflect.ValueOf(value))
+				} else {
+					targetValue.Set(reflect.ValueOf(value))
+				}
+			} else {
+				return err
+			}
+		}
+	default:
+		correspondingValue := targetValue
+		for correspondingValue.Kind() == reflect.Pointer {
+			correspondingValue = targetValue.Elem()
+		}
+
+		err := rg.checkChild(correspondingValue, res, metadata)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rg *ResourceGraph) callCreate(targetValue reflect.Value, metadata any) error {
+	method := targetValue.MethodByName("Create")
+	if method.IsValid() {
+		var callArgs []reflect.Value
+		callArgs = append(callArgs, reflect.ValueOf(rg))
+		params := reflect.New(method.Type().In(1)).Interface()
+		decoder := GetMapDecoder(params)
+		err := decoder.Decode(metadata)
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("error decoding the following type %s", reflect.New(method.Type().In(1)).Type().String()))
+		}
+		callArgs = append(callArgs, reflect.ValueOf(params).Elem())
+		eval := method.Call(callArgs)
+		if eval[0].IsNil() {
+			return nil
+		} else {
+			err, ok := eval[0].Interface().(error)
+			if !ok {
+				return fmt.Errorf("return type should be an error")
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (rg *ResourceGraph) checkChild(child reflect.Value, res Resource, metadata any) error {
+	var merr multierr.Error
```

Comment:
Are we on go 1.20? Maybe not worth it now, but https://go.dev/doc/go1.20#errors may make more sense than multierrors when we wanna convert

Reply:
i believe we are now, can move to the go errors

Reply:
Yeah, I think we can but I'd prefer we file that as tech debt and take it on later (or at least, deliberately and separate from this work)

Reply:
i cut a ticket for it so we dont forget, but going to address everything to unblock all the next tasks first